crates/hir-expand/src/fixup.rs

```diff
@@ -194,7 +193,105 @@ pub(crate) fn fixup_syntax(node: &SyntaxNode) -> SyntaxFixups {
+                            SyntheticToken {
+                                kind: SyntaxKind::L_CURLY,
+                                text: "{".into(),
+                                range: end_range,
+                                id: EMPTY_ID,
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::UNDERSCORE,
+                                text: "_".into(),
+                                range: end_range,
+                                id: EMPTY_ID
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::EQ,
+                                text: "=".into(),
+                                range: end_range,
+                                id: EMPTY_ID
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::R_ANGLE,
+                                text: ">".into(),
+                                range: end_range,
+                                id: EMPTY_ID
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::L_CURLY,
+                                text: "{".into(),
+                                range: end_range,
+                                id: EMPTY_ID,
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::R_CURLY,
+                                text: "}".into(),
+                                range: end_range,
+                                id: EMPTY_ID,
+                            },
+                            SyntheticToken {
+                                kind: SyntaxKind::R_CURLY,
+                                text: "}".into(),
+                                range: end_range,
+                                id: EMPTY_ID,
+                            },
+                        ]);
+                    }
+                },
+                ast::ForExpr(it) => {
+                    let for_token = match it.for_token() {
+                        Some(token) => token,
+                        None => continue
+                    };
+
+                    let [pat, in_token, iter] = [
+                        (SyntaxKind::UNDERSCORE, "_"),
+                        (SyntaxKind::IN_KW, "in"),
+                        (SyntaxKind::IDENT, "__ra_fixup")
+                    ].map(|(kind, text)| SyntheticToken { kind, text: text.into(), range: end_range, id: EMPTY_ID});
+
+                    if it.pat().is_none() && it.in_token().is_none() && it.iterable().is_none() {
+                        append.insert(for_token.into(), vec![pat, in_token, iter]);
+                    }
+
+                    // Tricky: add logic to add in just a pattern or iterable if not all
+                    // the pieces are missing
```

Comment:
for the missing pattern case we can just insert after the `for `keyword's text range

Reply:
I'm having some trouble with this, here's the relevant snippet (unindented a little):
```rust
        let [pat, in_token, iter] = [
            (SyntaxKind::UNDERSCORE, "_"),
            (SyntaxKind::IN_KW, "in"),
            (SyntaxKind::IDENT, "__ra_fixup")
        ].map(|(kind, text)| SyntheticToken { kind, text: text.into(), range: end_range, id: EMPTY_ID});

        if it.pat().is_none() && it.in_token().is_none() && it.iterable().is_none() {
            append.insert(for_token.into(), vec![pat, in_token, iter]);
        // does something funky -- see test case for_no_pat
        } else if it.pat().is_none() { 
            append.insert(for_token.into(), vec![pat]);
        } 

```
I also tried adding some other conditions like `it.in_token().is_some()` and `it.in_token().is_some() || it.iterable().is_some()` and `it.in_token().is_some() && it.iterable().is_some()` just in case the parser is getting confused. These don't have any effect though.

Here's the test output
```rust
    #[test] // Produces valid but incorrect code
    fn for_no_pat() {
        check(
            r#"
fn foo() {
    for in qux {

    }
}
"#,
            expect![[r#"
fn foo () {__ra_fixup}
"#]],
        )
    }
```

Reply:
the parser seems to wrap the `in` token in an error node which is why `in_token()` returns `None` here. That might be fixable by making the parser recover better in this situation but I'd have to look into that later.

Reply:
(An aside r-a has a `Show Syntax Tree`) comment that shows you the syntax tree of a file / your selection you can use that to debug these kinds of syntax things quickly. Accessible in VSCode via ctrl + shift + p and typing the command

Reply:
Ok, should I commit what I have now (for loop with no body/no iteration/both)? I can leave a note in the code about this.

Reply:
Thanks for the tip! I see the error node now.